File: User/Cocoa/AquaticPrime.m
//
// AquaticPrime.m
// AquaticPrime Framework
//
// Copyright (c) 2005, Lucas Newman
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without modification,
// are permitted provided that the following conditions are met:
//	¥Redistributions of source code must retain the above copyright notice,
//	 this list of conditions and the following disclaimer.
//	¥Redistributions in binary form must reproduce the above copyright notice,
//	 this list of conditions and the following disclaimer in the documentation and/or
//	 other materials provided with the distribution.
//	¥Neither the name of Aquatic nor the names of its contributors may be used to 
//	 endorse or promote products derived from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
// IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
// FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL 
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, 
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER 
// IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT 
// OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#import "AquaticPrime.h"

@implementation AquaticPrime

- (id)init
{
	return [self initWithKey:nil privateKey:nil];
}

- (id)initWithKey:(NSString *)key
{	
	return [self initWithKey:key privateKey:nil];
}

- (id)initWithKey:(NSString *)key privateKey:(NSString *)privateKey
{
	ERR_load_crypto_strings();
	
	if (![super init])
		return nil;
	
	aqError = [[NSString alloc] init];
	blacklist = [[NSArray alloc] init];
	hash = [[NSString alloc] init];
    rsaKey = nil;
	
	[self setKey:key privateKey:privateKey];
	
	return self;
}

- (void)dealloc
{	
	ERR_free_strings();
	
	if (rsaKey)
		RSA_free(rsaKey);

	[blacklist release];
	[aqError release];
	[hash release];
	
	[super dealloc];
}

+ (id)aquaticPrimeWithKey:(NSString *)key privateKey:(NSString *)privateKey
{
	return [[[AquaticPrime alloc] initWithKey:key privateKey:privateKey] autorelease];
}

+ (id)aquaticPrimeWithKey:(NSString *)key
{
	return [[[AquaticPrime alloc] initWithKey:key privateKey:nil] autorelease];
}

- (BOOL)setKey:(NSString *)key 
{
	return [self setKey:key privateKey:nil];
}

- (BOOL)setKey:(NSString *)key privateKey:(NSString *)privateKey
{
	// Must have public modulus, private key is optional
	if (!key || [key isEqualToString:@""]) {
		[self _setError:@"Empty public key parameter"];
		return NO;
	}
	
	if (rsaKey)
		RSA_free(rsaKey);
		
	rsaKey = RSA_new();
	
	// We are using the constant public exponent e = 3
	BN_dec2bn(&rsaKey->e, "3");
	
	// Determine if we have hex or decimal values
	int result;
	if ([[key lowercaseString] hasPrefix:@"0x"])
		result = BN_hex2bn(&rsaKey->n, (const char *)[[key substringFromIndex:2] UTF8String]);
	else
		result = BN_dec2bn(&rsaKey->n, (const char *)[key UTF8String]);
		
	if (!result) {
		[self _setError:[NSString stringWithUTF8String:(char*)ERR_error_string(ERR_get_error(), NULL)]];
		return NO;
	}
	
	// Do the private portion if it exists
	if (privateKey && ![privateKey isEqualToString:@""]) {
		if ([[privateKey lowercaseString] hasPrefix:@"0x"])
			result = BN_hex2bn(&rsaKey->d, (const char *)[[privateKey substringFromIndex:2] UTF8String]);
		else
			result = BN_dec2bn(&rsaKey->d, (const char *)[privateKey UTF8String]);
			
		if (!result) {
			[self _setError:[NSString stringWithUTF8String:(char*)ERR_error_string(ERR_get_error(), NULL)]];
			return NO;
		}
	}
	
	return YES;
}

- (NSString *)key
{
	if (!rsaKey || !rsaKey->n)
		return nil;
	
	char *cString = BN_bn2hex(rsaKey->n);
	
	NSString *nString = [[NSString alloc] initWithUTF8String:cString];
	OPENSSL_free(cString);
	
	return nString;
}

- (NSString *)privateKey
{	
	if (!rsaKey || !rsaKey->d)
		return nil;
	
	char *cString = BN_bn2hex(rsaKey->d);
	
	NSString *dString = [[NSString alloc] initWithUTF8String:cString];
	OPENSSL_free(cString);
	
	return dString;
}

- (void)setHash:(NSString *)newHash
{
	[hash release];
	hash = [newHash retain];
}

- (NSString *)hash
{
	return hash;
}

#pragma mark Blacklisting

// This array should contain a list of NSStrings representing hexadecimal hashcodes for blacklisted licenses
- (void)setBlacklist:(NSArray*)hashArray
{
	[blacklist release];
	blacklist = [hashArray retain];
}

#pragma mark Signing

- (NSData*)licenseDataForDictionary:(NSDictionary*)dict
{	
	// Make sure we have a good key
	if (!rsaKey || !rsaKey->n || !rsaKey->d) {
		[self _setError:@"RSA key is invalid"];
		return nil;
	}
	
	// Grab all values from the dictionary
	NSMutableArray *keyArray = [NSMutableArray arrayWithArray:[dict allKeys]];
	NSMutableData *dictData = [NSMutableData data];
	
	// Sort the keys so we always have a uniform order
	[keyArray sortUsingSelector:@selector(caseInsensitiveCompare:)];
	
	int i;
	for (i = 0; i < [keyArray count]; i++)
	{
		id curValue = [dict objectForKey:[keyArray objectAtIndex:i]];
		char *desc = (char *)[[curValue description] UTF8String];
		// We use strlen instead of [string length] so we can get all the bytes of accented characters
		[dictData appendBytes:desc length:strlen(desc)];
	}
	
	// Hash the data
	unsigned char digest[20];
	SHA1([dictData bytes], [dictData length], digest);
	
	// Create the signature from 20 byte hash
	int rsaLength = RSA_size(rsaKey);
	unsigned char *signature = (unsigned char*)malloc(rsaLength);
	int bytes = RSA_private_encrypt(20, digest, signature, rsaKey, RSA_PKCS1_PADDING);
	
	if (bytes == -1) {
		[self _setError:[NSString stringWithUTF8String:(char*)ERR_error_string(ERR_get_error(), NULL)]];
		return nil;
	}
	
	// Create the license dictionary
	NSMutableDictionary *licenseDict = [NSMutableDictionary dictionaryWithDictionary:dict];
	[licenseDict setObject:[NSData dataWithBytes:signature length:bytes]  forKey:@"Signature"];
	
	// Create the data from the dictionary
	NSString *error;
	NSData *licenseFile = [[NSPropertyListSerialization dataFromPropertyList:licenseDict 
														format:kCFPropertyListXMLFormat_v1_0 
														errorDescription:&error] retain];
	
	if (!licenseFile) {
		[self _setError:error];
		return nil;
	}
	
	return licenseFile;
}

- (BOOL)writeLicenseFileForDictionary:(NSDictionary*)dict toPath:(NSString *)path
{
	NSData *licenseFile = [self licenseDataForDictionary:dict];
	
	if (!licenseFile)
		return NO;
	
	return [licenseFile writeToFile:path atomically:YES];
}

// This method only logs errors on developer problems, so don't expect to grab an error message if it's just an invalid license
- (NSDictionary*)dictionaryForLicenseData:(NSData *)data
{	
	// Make sure public key is set up
	if (!rsaKey || !rsaKey->n) {
		[self _setError:@"RSA key is invalid"];
		return nil;
	}

	// Create a dictionary from the data
	NSPropertyListFormat format;
	NSString *error;
	NSMutableDictionary *licenseDict = [NSPropertyListSerialization propertyListFromData:data mutabilityOption:NSPropertyListMutableContainersAndLeaves format:&format errorDescription:&error];
	if (![licenseDict isKindOfClass:[NSMutableDictionary class]] || error)
		return nil;
		
	NSData *signature = [licenseDict objectForKey:@"Signature"];
	if (!signature)
		return nil;
	
	// Decrypt the signature - should get 20 bytes back
	unsigned char checkDigest[20];
	if (RSA_public_decrypt([signature length], [signature bytes], checkDigest, rsaKey, RSA_PKCS1_PADDING) != 20)
		return nil;
	
	// Make sure the license hash isn't on the blacklist
	NSMutableString *hashCheck = [NSMutableString string];
	int hashIndex;
	for (hashIndex = 0; hashIndex < 20; hashIndex++)
		[hashCheck appendFormat:@"%02x", checkDigest[hashIndex]];
	
	// Store the license hash in case we need it later
	[self setHash:hashCheck];
	
	if (blacklist && [blacklist containsObject:hashCheck])
			return nil;
	
	// Remove the signature element
	[licenseDict removeObjectForKey:@"Signature"];
	
	// Grab all values from the dictionary
	NSMutableArray *keyArray = [NSMutableArray arrayWithArray:[licenseDict allKeys]];
	NSMutableData *dictData = [NSMutableData data];
	
	// Sort the keys so we always have a uniform order
	[keyArray sortUsingSelector:@selector(caseInsensitiveCompare:)];
	
	int objectIndex;
	for (objectIndex = 0; objectIndex < [keyArray count]; objectIndex++)
	{
		id currentValue = [licenseDict objectForKey:[keyArray objectAtIndex:objectIndex]];
		char *description = (char *)[[currentValue description] UTF8String];
		// We use strlen instead of [string length] so we can get all the bytes of accented characters
		[dictData appendBytes:description length:strlen(description)];
	}
	
	// Hash the data
	unsigned char digest[20];
	SHA1([dictData bytes], [dictData length], digest);
	
	// Check if the signature is a match	
	int checkIndex;
	for (checkIndex = 0; checkIndex < 20; checkIndex++) {
		if (checkDigest[checkIndex] ^ digest[checkIndex])
			return nil;
	}
	
	return [NSDictionary dictionaryWithDictionary:licenseDict];
}

- (NSDictionary*)dictionaryForLicenseFile:(NSString *)path
{
	NSData *licenseFile = [NSData dataWithContentsOfFile:path];
	
	if (!licenseFile)
		return nil;
	
	return [self dictionaryForLicenseData:licenseFile];
}

- (BOOL)verifyLicenseData:(NSData *)data
{
	if ([self dictionaryForLicenseData:data])
		return YES;
	else
		return NO;
}

- (BOOL)verifyLicenseFile:(NSString *)path
{
	NSData *data = [NSData dataWithContentsOfFile:path];
	return [self verifyLicenseData:data];
}

#pragma mark Error Handling

- (NSString*)getLastError
{
	return aqError;
}

@end

@implementation AquaticPrime (Private)

- (void)_setError:(NSString *)err
{
	[aqError release];
	aqError = [err retain];
}

@end

File: Source/STL/AquaticPrime.cpp
//
// AquaticPrime.cpp
// AquaticPrime STL Implementation
//
// Copyright (c) 2005, Lucas Newman
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without modification,
// are permitted provided that the following conditions are met:
//	¥Redistributions of source code must retain the above copyright notice,
//	 this list of conditions and the following disclaimer.
//	¥Redistributions in binary form must reproduce the above copyright notice,
//	 this list of conditions and the following disclaimer in the documentation and/or
//	 other materials provided with the distribution.
//	¥Neither the name of the Aquatic nor the names of its contributors may be used to 
//	 endorse or promote products derived from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
// IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
// FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL 
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, 
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER 
// IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT 
// OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#include "AquaticPrime.h"
#include "tinyxml.h"
#include <stdarg.h>
#include <b64/b64.h>

static RSA *rsaKey;
static std::string hash;
static std::list<std::string> blacklist;

// utilities
inline char ToLower(char in)
{
	return (char)tolower((int)in);
}

const char* CreateCString(std::string output, ...)
{
	static char text[256];
	va_list	ap;

	va_start(ap, output);								// Parses The String For Variables
		vsprintf(text, output.c_str(), ap);				// And Converts Symbols To Actual Numbers
	va_end(ap);					

	return (const char*)text;
}

bool APSetKey(std::string key)
{
	hash = std::string("");
	
	// Create a new key
    rsaKey = RSA_new();
    
    // Public exponent is always 3
	BN_hex2bn(&rsaKey->e, "3");
	
	std::string mutableKey = key;
    
    // Determine if we have a hex or decimal key
	std::transform(mutableKey.begin(), mutableKey.end(), mutableKey.begin(), ToLower); // make mutableKey lowercase
	if(std::string(mutableKey, 0, 2) == "0x")
	{
		mutableKey = std::string(mutableKey, 2, mutableKey.length());
		BN_hex2bn(&rsaKey->n, mutableKey.c_str());
	}
	else 
	{
		BN_dec2bn(&rsaKey->n, mutableKey.c_str());
	}
	
	return true;
}

std::string APHash(void)
{
	return hash;
}

void APSetHash(std::string newHash)
{
	hash = newHash;
}

// Set the entire blacklist array, removing any existing entries
void APSetBlacklist(std::list<std::string> hashArray)
{
	blacklist = hashArray;
}

void APBlacklistAdd(std::string blacklistEntry)
{
	blacklist.push_back(blacklistEntry);
}

std::map<std::string, std::string> APCreateDictionaryForLicenseData(std::map<std::string, std::string> data)
{
	if (!rsaKey->n || !rsaKey->e)
	{
		std::map<std::string, std::string> empty;
		printf("0\n");
		return empty;
	}
	
	// Load the signature
	unsigned char sigBytes[128];
	std::map<std::string, std::string>::iterator signature = data.find("Signature");

    if(signature == data.end())
	{
		std::map<std::string, std::string> empty;
//		printf("1\n");
		return empty;
	}
	else 
	{
		int returnVal = b64::b64_decode(data["Signature"].c_str(), data["Signature"].length(), sigBytes, 129);	
		
		if(returnVal == 0)
		{
			std::map<std::string, std::string> empty;
	//		printf("1.5\n");
			return empty;
		}
		
		data.erase(signature);
	}
	
	// Decrypt the signature
	unsigned char checkDigest[128] = {0};
	if (RSA_public_decrypt(128, sigBytes, checkDigest, rsaKey, RSA_PKCS1_PADDING) != SHA_DIGEST_LENGTH)
    {
		std::map<std::string, std::string> empty;
//		printf("2\n");
		return empty;
	}

	// Get the license hash
	std::string hashCheck;
	int hashIndex;
	for (hashIndex = 0; hashIndex < SHA_DIGEST_LENGTH; hashIndex++)
		hashCheck += CreateCString("%02x", checkDigest[hashIndex]);
	APSetHash(hashCheck);
	
	if (blacklist.size() > 0)
	{
		// $$ is this right?
		for(std::list<std::string>::iterator b = blacklist.begin(); b != blacklist.end(); ++b)
		{
			if(data.find((*b)) != data.end())
			{
				std::map<std::string, std::string> empty;
//				printf("3\n");
				return empty;
			}
		}
	}
	
	// Get the number of elements
	int count = data.size();
	// Load the keys and build up the key array
//	std::list<std::string> keyArray;
	std::string keys[count];
	
	int counter = 0;
	for(std::map<std::string, std::string>::iterator d = data.begin(); d != data.end(); ++d)
	{
		keys[counter] = (*d).first;
		++counter;
	}
	
	// Sort the array ( $$ why?  for cleanliness reasons? )
//	int context = kCFCompareCaseInsensitive;
//	CFArraySortValues(keyArray, CFRangeMake(0, count), (CFComparatorFunction)CFStringCompare, &context);
	
	// Setup up the hash context
	SHA_CTX ctx;
	SHA1_Init(&ctx);
	// Convert into UTF8 strings
	for(int i = 0; i < count; i++)
	{
		std::string key = keys[i]; // $$ convert this to keyArray later
		std::string value = data[key];

		// Account for the null terminator
		SHA1_Update(&ctx, value.c_str(), strlen(value.c_str()));
	}
	unsigned char digest[SHA_DIGEST_LENGTH];
	SHA1_Final(digest, &ctx);
	
	// Check if the signature is a match	
	for (int i = 0; i < SHA_DIGEST_LENGTH; i++) 
	{
		if (checkDigest[i] ^ digest[i]) 
		{
			std::map<std::string, std::string> empty;
//			printf("4\n");
			return empty;
        }
	}

	// If it's a match, we return the dictionary; otherwise, we never reach this
	return data;
}

std::map<std::string, std::string> APCreateDictionaryForLicenseFile(std::string path)
{
	std::map<std::string, std::string> xmlData;
	
	TiXmlNode *node = 0;
	TiXmlDocument licenseFile(path.c_str());
	licenseFile.LoadFile();
	
	node = licenseFile.FirstChild("plist");
	if(node == NULL) return xmlData;
	node = node->FirstChild("dict");
	if(node == NULL) return xmlData;
	
	do
	{
		// <dict>
		if(std::string(node->ToElement()->Value()) == std::string("dict"))
		{
			std::string key, data;
			TiXmlNode *innerNode = node->FirstChild();
			while(innerNode != NULL)
			{				
				// <key>
				if(std::string(innerNode->ToElement()->Value()) == std::string("key"))
				{
					key = innerNode->ToElement()->FirstChild()->Value();
//					printf("key %s\n", key.c_str());
				}
				// <string>
				else if(std::string(innerNode->ToElement()->Value()) == std::string("string"))
				{
					data = innerNode->ToElement()->FirstChild()->Value();
					xmlData[key] = data;
//					printf("string %s %s\n", key.c_str(), data.c_str());
				}
				// <data>
				else if(std::string(innerNode->ToElement()->Value()) == std::string("data"))
				{
					data = innerNode->ToElement()->FirstChild()->Value();
					
					if(key == "Signature") // get rid of any spaces
					{
						std::vector<std::string::iterator> spaces;
						for(std::string::iterator d = data.begin(); d != data.end(); ++d)
						{
							if((*d) == ' ')
								spaces.push_back(d);
						}

						for(uint s=0; s < spaces.size(); ++s)
							data.erase(spaces[s] - (s));
					}
					
					xmlData[key] = data;
//					printf("data %s %s\n", key.c_str(), data.c_str());
				}
				
				innerNode = innerNode->NextSibling();
			}
		}
		
		node = node->NextSibling();
	} while(node != NULL);

	std::map<std::string, std::string> licenseDictionary = APCreateDictionaryForLicenseData(xmlData);

	return licenseDictionary;
}

bool APVerifyLicenseData(std::map<std::string, std::string> data)
{
	std::map<std::string, std::string> licenseDictionary = APCreateDictionaryForLicenseData(data);
	
	if (licenseDictionary.size() > 0)
		return true;
	else
		return false;
}

bool APVerifyLicenseFile(std::string path)
{
	std::map<std::string, std::string> licenseDictionary = APCreateDictionaryForLicenseFile(path);

	if (licenseDictionary.size() > 0)
		return true;
	else
		return false;
}
